corehq/sql_proxy_accessors/README.md

```diff
@@ -1,6 +1,7 @@
 # Sharding postgresql
 
-We use [PL/Proxy](https://plproxy.github.io/) for sharding.
+We use [PL/Proxy](https://plproxy.github.io/) for sharding, although as of September 2017
+are in the process of phasing it out in favor of sharding in the application layer.
```

Comment:
Not very excited about this. One unfortunate consequence of sharding in the application layer is that queries to multiple shards will almost certainly be executed serially (because of the GIL) rather than in parallel unless we do something like write a C extension for sharded queries (non-trivial). I'm not sure if PL/Proxy does that, but I think [Citus has features like parallel execution](https://docs.citusdata.com/en/v6.2/dist_tables/querying.html#querying-distributed-tables-sql).

Reply:
Hmm, this seems worth talking to @snopoke about. I was under the impression the plan was to move everything to the application layer to make future operations like shard migrations easier to manage, but not sure if parallellization was discussed/considered. feel free to hit me up if there's a meeting I should join to listen in on or if I should jump on slack.

Reply:
(will follow up on email)

Reply:
> I'm not sure if PL/Proxy does that, but Citus has features like parallel execution.

according to https://wiki.postgresql.org/wiki/PL/Proxy it does execute in parallel

Reply:
Any suggestions @millerdev?

Reply:
I know @snopoke and I had discussed this briefly in the past, and I came away thinking the work he was doing looked like good architectural cleanup without many far-reaching effects (that is, easy to switch back if we wanted). Maybe I'm forgetting what I reviewed before, but this work that Cory is doing looks more invasive.

I have had reservations about moving sharding logic from PL/Proxy into Python ever since I first heard about it, and I'm still not on board with the idea. I have not devoted a lot of time to making a case so far because I haven't been working on it directly and it's felt like I'd have to drop other things to find enough time to make a solid case. I'd like to do that if I can. The place I'd like to start is analyzing our current performance bottleneck(s) to see if we can really pin the blame on PL/Proxy and answer the question will moving sharding logic into Python improve performance? (my gut feeling says no way, but that's all I've got right now.) I'm going to make some time today to look into this a bit more.